### PR TITLE
proxy should be used when making curl request #7658

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: config/defaults.inc.php [hunk 1]
```diff
@@ -795,6 +795,11 @@
 // PHP code about the location of asset files in filesystem
 $config['assets_dir'] = '';
 
+// http(s) proxy to use for fetching remote resources using curl
+$config['http_client'] = array(
+  'timeout' => 10,
+  'proxy' => 'tcp://localhost:8125',
```
Comment (alecpl) [inline config/defaults.inc.php:801]: Please, comment out the array content.

File: program/steps/utils/modcss.inc [hunk 2]
```diff
@@ -47,17 +47,11 @@ if (ini_get('allow_url_fopen')) {
     $headers = implode("\n", (array) $http_response_header);
 }
 else if (function_exists('curl_init')) {
```
Comment (alecpl) [inline program/steps/utils/modcss.inc:49]: I would remove that `else if` and the whole `if` before.

**Review by alecpl — CHANGES_REQUESTED** [review]
Please, resolve conflict. Also, it would be good to make use of the new proxy setting and guzzle in rcube_spellchecker_googie. Could you do this?

File: config/defaults.inc.php (hunk 1 above)
Comment (alecpl) [reply config/defaults.inc.php:801]: Also the option description needs update

Comment (ata-sql) [reply config/defaults.inc.php:801]: Sure I will change guzzle in rcube_spellchecker_googie also.

**Comment by ata-sql** [comment]
OK, changed according to your suggestions

File: program/include/rcmail_utils.php [hunk 3]
```diff
@@ -18,6 +18,8 @@
  +-----------------------------------------------------------------------+
 */
 
+use GuzzleHttp\Client;
```
Comment (alecpl) [inline program/include/rcmail_utils.php:21]: It's used in only one place. I would remove the use statement (and add a backslash below).

File: program/lib/Roundcube/spellchecker/googie.php [hunk 4]
```diff
@@ -55,24 +54,22 @@ function languages()
...
 
         // spell check uri is configured
         $url = rcube::get_instance()->config->get('spellcheck_uri');
 
         if ($url) {
             $a_uri = parse_url($url);
-            $ssl   = ($a_uri['scheme'] == 'https' || $a_uri['scheme'] == 'ssl');
-            $port  = $a_uri['port'] ? $a_uri['port'] : ($ssl ? 443 : 80);
-            $host  = ($ssl ? 'ssl://' : '') . $a_uri['host'];
+            $host  =  $a_uri['host'];
```
Comment (alecpl) [inline program/lib/Roundcube/spellchecker/googie.php:68]: Redundant space.

File: program/lib/Roundcube/spellchecker/googie.php [hunk 5]
```diff
@@ -55,24 +54,22 @@ function languages()
...
+            return $this->matches = $matches;
         }
 
         // spell check uri is configured
         $url = rcube::get_instance()->config->get('spellcheck_uri');
 
         if ($url) {
             $a_uri = parse_url($url);
-            $ssl   = ($a_uri['scheme'] == 'https' || $a_uri['scheme'] == 'ssl');
-            $port  = $a_uri['port'] ? $a_uri['port'] : ($ssl ? 443 : 80);
```
Comment (alecpl) [inline program/lib/Roundcube/spellchecker/googie.php:69]: What if the port is not 443 nor 80? It looks to me as BC break. Configured port needs to be taken into account. I don't really know Guzzle, is this a problem?

File: program/lib/Roundcube/spellchecker/googie.php [hunk 6]
```diff
@@ -83,47 +80,36 @@ function check($text)
...
+        $response = $client->send($request);
 
-        // parse HTTP response
-        if (preg_match('!^HTTP/1.\d (\d+)(.+)!', $store, $m)) {
-            $http_status = $m[1];
-            if ($http_status != '200') {
-                $this->error = 'HTTP ' . $m[1] . rtrim($m[2]);
-            }
+        if (empty($response)) {
+          $this->error = "Empty result from spelling engine";
```
Comment (alecpl) [inline program/lib/Roundcube/spellchecker/googie.php:91]: 4 spaces padding, please.

File: program/steps/utils/modcss.inc [hunk 7]
```diff
@@ -31,33 +31,12 @@ if (!preg_match('~^(https?)://~i', $realurl, $matches)) {
...
-    curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, 15);
-    curl_setopt($curl, CURLOPT_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS);
-    curl_setopt($curl, CURLOPT_ENCODING, '');
-    curl_setopt($curl, CURLOPT_HEADER, true);
-    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
-    $data = curl_exec($curl);
-
-    if ($data !== false) {
-        list($headers, $source) = explode("\r\n\r\n", $data, 2);
+if (function_exists('curl_init')) {
```
Comment (alecpl) [inline program/steps/utils/modcss.inc:34]: Either we remove the condition completely or test for 'allow_url_fopen' too, as this is the method that Guzzle could use if there's no curl. I think we can remove this.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/lib/Roundcube/spellchecker/googie.php (hunk 5 above)
Comment (ata-sql) [reply program/lib/Roundcube/spellchecker/googie.php:69]: It's not a problem - you provide URI - if it uses non-standard port, then URI will deliver it:
`$config['spellcheck_uri'] = 'protocol://host.domain:port';`
I don't recognize it as BC break.

File: program/lib/Roundcube/spellchecker/googie.php (hunk 4 above)
Comment (ata-sql) [reply program/lib/Roundcube/spellchecker/googie.php:68]: fixed

File: program/include/rcmail_utils.php (hunk 3 above)
Comment (ata-sql) [reply program/include/rcmail_utils.php:21]: fixed

File: program/lib/Roundcube/spellchecker/googie.php (hunk 6 above)
Comment (ata-sql) [reply program/lib/Roundcube/spellchecker/googie.php:91]: Fixed. Sorry for that - it's my IDE setting.

File: program/steps/utils/modcss.inc (hunk 7 above)
Comment (ata-sql) [reply program/steps/utils/modcss.inc]: removed

File: .gitignore [hunk 8]
```diff
@@ -25,3 +25,5 @@ program/js/tinymce/
 # skin customization files
 skins/elastic/styles/_styles.less
 skins/elastic/styles/_variables.less
+
+.idea
```
Comment (alecpl) [inline .gitignore:29]: This also should go away.

File: program/lib/Roundcube/spellchecker/googie.php (hunk 5 above)
Comment (alecpl) [reply program/lib/Roundcube/spellchecker/googie.php:69]: But you pass only $host to Guzzle, port is ignored.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 8 above)
Comment (ata-sql) [reply .gitignore:29]: Why? You don't use IntelliJ or other IDE? It hurts something?

**Comment by ata-sql** [comment]
bleeh, just realized, that I was using Guzzle 3.x version (Centos 8) and newer versions are not BC.
Will rewrite it for  Guzzle as declared in composer ("guzzlehttp/guzzle": "^6.5.5").

**Review by ghost — COMMENTED** [review]
> bleeh, just realized, that I was using Guzzle 3.x version (Centos 8) and newer versions are not BC.
> Will rewrite it for Guzzle as declared in composer ("guzzlehttp/guzzle": "^6.5.5").

OK, fixed that.

File: .gitignore (hunk 8 above)
Comment (ata-sql) [reply .gitignore]: changed

**Comment by ata-sql** [comment]
Sorry, anything more needed to merge that PR?

**Comment by alecpl** [comment]
Please, move the .idea folder to a separate PR if you really want it to be considered for inclusion. This way it will be easier to review and accept this request.

I'm in the middle of some refactoring, so it will have to wait some time, but I'll get back to you.

**Comment by ata-sql** [comment]
gitignore changes removed from that PR.

**Comment by alecpl** [comment]
The refactoring has been merged into master. Please, rebase.

**Comment by ata-sql** [comment]
> The refactoring has been merged into master. Please, rebase.

rebased

**Comment by alecpl** [comment]
It does not look right. You should update `program/actions/utils/modcss.php`, not create modcss.inc

**Comment by ata-sql** [comment]
> It does not look right. You should update `program/actions/utils/modcss.php`, not create modcss.inc

Fixed. It wasn't conflicting and you refactor was quite big, so I didn't notice that.
Btw. you omit 2 occurences for modcss.inc after refactor:
image

**Review by alecpl — APPROVED** [review]

**Pull request merged**